tools/content/fetch_messages.dart

```diff
@@ -0,0 +1,223 @@
+  final email = parsedConfig.get('api', 'email') as String;
+  final apiKey = parsedConfig.get('api', 'key') as String;
+  final site = Uri.parse(parsedConfig.get('api', 'site') as String);
+
+  final outputDirStr = parsedArguments['corpus-dir'] as String?;
+  final fetchNewer = parsedArguments['fetch-newer'] as bool;
+  int? anchorMessageId;
+  IOSink output = stdout;
+  if (outputDirStr != null) {
+    final outputDir = Directory(outputDirStr);
+    outputDir.createSync(recursive: true);
+    final outputFile = File('$outputDirStr/${site.host}.jsonl');
+    if (!outputFile.existsSync()) outputFile.createSync();
+    // Look for the known newest/oldest message so that we can continue
+    // fetching from where we left off.
+    await for (final message in readMessagesFromJsonl(outputFile)) {
+      anchorMessageId ??= message.id;
+      // Newer Zulip messages have higher message IDs.
+      anchorMessageId = (fetchNewer ? max : min)(message.id, anchorMessageId);
+    }
+    output = outputFile.openWrite(mode: FileMode.writeOnlyAppend);
+  }
+
+  final client = Client();
+  final authHeader = 'Basic ${base64Encode(utf8.encode('$email:$apiKey'))}';
+
+  // These are working constants chosen abitrarily.
+  const batchSize = 5000;
+  const maxRetries = 10;
+  const fetchInterval = Duration(seconds: 5);
+
+  int retries = 0;
+  BackoffMachine? backoff;
+
+  while (true) {
+    // This loops until there is no message fetched in an iteration.
+    final _GetMessagesResult result;
+    try {
+      result = await _getMessages(client, realmUrl: site,
+        authHeader: authHeader,
+        anchorMessageId: anchorMessageId,
+        numBefore: (!fetchNewer) ? batchSize : 0,
+        numAfter: (fetchNewer) ? batchSize : 0,
+      );
+    } catch (e) {
+      // We could have more fine-grained error handling and avoid retrying on
+      // non-network-related failures, but that's not necessary.
+      if (retries >= maxRetries) {
+        rethrow;
+      }
+      retries++;
+      await (backoff ??= BackoffMachine()).wait();
+      continue;
+    }
+
+    final messageEntries = result.messages.map(MessageEntry.fromJson);
+    if (messageEntries.isEmpty) {
+      // Sanity check to ensure that the server agrees
+      // there is no more messages to fetch.
+      if (fetchNewer) assert(result.foundNewest);
+      if (!fetchNewer) assert(result.foundOldest);
+      break;
+    }
+
+    // Find and use the newest/oldest message as the next message fetch anchor.
+    anchorMessageId = messageEntries.map((x) => x.id).reduce(fetchNewer ? max : min);
+    messageEntries.map(jsonEncode).forEach((json) => output.writeln(json));
+
+    // This I/O operation could fail, but crashing is fine here.
+    final flushFuture = output.flush();
+    // Make sure the delay happens concurrently to the flush.
+    await Future<void>.delayed(fetchInterval);
+    await flushFuture;
+    backoff = null;
+  }
+  exit(0);
+}
```

Comment:
Does this `exit` call differ from simply falling off the end of `main` here and returning?